framework/aster-frame/src/mm/io.rs

```diff
@@ -168,28 +176,133 @@ impl_vmio_pointer!(&mut T, "(**self)");
+pub struct KernelSpace;
+
+impl private::Sealed for UserSpace {}
+impl private::Sealed for KernelSpace {}
+
+/// Copies `len` bytes from `src` to `dst`.
+/// Moves `src` and `dst` to the end of copied range.
+///
+/// Returns the number of successfully copied bytes.
+///
+/// # Safety
+///
+/// Mappings of virtual memory range [`src`..`src` + len] and [`dst`..`dst` + len]
+/// must be pinned. If one of the memory represent typed memory, these two virtual
+/// memory ranges and their corresponding physical pages should _not_ overlap.
+///
+/// Operation on typed memory may be safe only if it is plain-old-data. Otherwise
+/// the safety requirements of [`core::ptr::copy`] should also be considered.
+unsafe fn copy_with_pinned_memmap(src: &mut *const u8, dst: &mut *mut u8, len: usize) -> usize {
+    core::ptr::copy(*src, *dst, len);
+    *src = src.add(len);
+    *dst = dst.add(len);
+
+    len
+}
+
+/// Copies `len` bytes from `src` to `dst`.
+/// Moves `src` and `dst` to the end of copied range.
+/// This function will early stop copying if encountering an unresolvable page fault.
+///
+/// Returns the number of successfully copied bytes.
+///
+/// # Safety
+///
+/// This method should only be used when one of [`src`..`src` + len] and [`dst`..`dst` + len]
+/// is in user space, and the other virtual memory range is in kernel space and is ensured to
+/// be accessible. In addition, users should ensure this function only be invoked when a suitable
+/// page table is activated.
+///
+/// The actual physical address of memory [`src`..`src` + len] and [`dst`..`dst` + len] should
+/// _not_ overlap if the kernel space memory represent typed memory.
+unsafe fn copy_may_page_fault(src: &mut *const u8, dst: &mut *mut u8, len: usize) -> usize {
+    let failed_bytes = copy_with_recovery(*dst, *src, len);
+    let copied_len = len - failed_bytes;
+    *src = src.add(copied_len);
+    *dst = dst.add(copied_len);
+
+    copied_len
+}
+
+/// `VmReader` is a reader for reading data from a contiguous range of memory.
+///
+/// The memory range read by `VmReader` can be in either kernel space or user space.
+/// When the operating range is in kernel space, the memory within that range
+/// is guaranteed to be valid.
+/// When the operating range is in user space, it is ensured that the page table of
+/// the process creating the `VmReader` is active for the duration of `'a`.
+///
+/// When perform reading with a `VmWriter`, if one of them represents typed memory,
+/// it can ensure that the reading range in this reader and writing range in the
+/// writer are not overlapped.
+///
+/// NOTE: The overlap mentioned above is at both the virtual address level
+/// and physical address level. There is not guarantee for the operation results
+/// of `VmReader` and `VmWriter` in overlapping untyped addresses, and it is
+/// the user's responsibility to handle this situation.
+pub struct VmReader<'a, Target: 'static + private::Sealed = KernelSpace> {
```

Comment:
In Rust std, sometimes the generic parameter is called `Target`, but that is because the type is a smarter pointer that refers to an object of `Target`. That is why calling the parameter `Target` is appropriate. But we normally would not say a reader or writer _targets_ the kernel or user space. Think of some better names. For example, `Space` or `AddrSpace` is better than `Target`.